### PR TITLE
Revision of the repo check workflows

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR refactors the spell and link checks, adds a workflow to check the formatting of C++ and Python files, and a check that all files have a copyright header. It creates a new workflow with various repo checks that replaces the old spell and link checker, and adds header and formatting checks. 

Spell checking for C++: checks all comments in header files as well as cpp files in the examples. Content between backticks (inline or as block) is treated as code and ignored. Strings are unchecked.
Spell checking in Python: checks all comments, doc comments, and strings in python files (source code and examples). Content between backticks (inline or as block) is treated as code and ignored. 
Markdown files: Code is ignored, link titles are checked.
Rst files: Processed like markdown files, with some additional rules to filter lines starting with two dots.

Not currently covered: Docs strings of the code generated by PyBind, since C++ strings are not spell checked. I think the best way to catch those is to simply spell check the html. I've added that check and I think it is configured decently, but it is currently  producing tons of errors due to minor fixes we need to make everywhere (such as properly marking references to class names as such, avoid abbreviations, etc.). The check is hence running but not enforced, meaning the build passes despite the spell check raising errors.

### CODE REVIEW

File: .github/workflows/config/spelling_allowlist.txt [hunk 1]
```diff
@@ -0,0 +1,78 @@
+NVIDIA
```
Comment (bettinaheim) [inline .github/workflows/config/spelling_allowlist.txt:1]: To be sorted in alphabetical order.

File: .github/workflows/config/spellcheck_config.yml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  - pyspelling.filters.context:
+      context_visible_first: true
+      escapes: '\\[\\`~]'
+      delimiters:
+      # Ignore multiline content between three or more backticks
+      - open: '(?s)^(?P<open>\s*`{3,})\S*\s*$'
+        close: '^(?P=open)$'
+      # Ignore content between inline backticks
+      - open: '(?P<open>`+)'
+        close: '(?P=open)'
```
Comment (schweitzpgi) [inline .github/workflows/config/spellcheck_config.yml:143]: So the backticks aren't markdown necessarily, but a way to defeat the spellchecker. What about `<code>` .. `</code>` blocks?

Also, can we just filter out any word that is written in camelCase or CamelCase without requiring backticks?

**Review by schweitzpgi — COMMENTED** [review]
LGTM

File: .github/workflows/config/spellcheck_config.yml (hunk 2 above)
Comment (bettinaheim) [reply .github/workflows/config/spellcheck_config.yml:143]: I think it is helpful if we can distinguish somehow what is code from what is "normal language". To make the spellchecker happy we could regex match any words that contain a capital letter that is not at the beginning of the word and ignore them. We can treat all matches as code and can at some point in the future also validate that documentation for this parameter/class name/whatever else actually exists. The reason that I am a bit less fond of automatic detection of camel casing is that you could always have a simple class name `foo` that is not distinguishable by looking for camel case. I hence think there necessarily always needs to be a way to explicitly mark words as code, and we should just use that everywhere. Whether that's backticks or xml doesn't matter to me, but we need to make sure doxygen or sphinx is properly recognizing it as code.

**Review by schweitzpgi — APPROVED** [review]
Approving to get the spell check CI unblocked. (There may still be a couple of things to address here?)

File: .github/workflows/config/spellcheck_config.yml (hunk 2 above)
Comment (schweitzpgi) [reply .github/workflows/config/spellcheck_config.yml:143]: OK. Yeah, it's a catch-22 sort of situation.

Comment (bettinaheim) [reply .github/workflows/config/spellcheck_config.yml:149]: Doxygen seems to be rather good at auto detecting references to code without any kind of marker. I hence added a filter for the spellcheck to ignore words that contains capital letters, underscores, or numbers. However, other things, such as e.g. when referring to a parameter name `idx` would need to be fenced with backticks.

**Comment by bettinaheim** [comment]
> Approving to get the spell check CI unblocked. (There may still be a couple of things to address here?)

I made the remaining updates. See also my reply on the other thread, and the updated description in the PR.

**Pull request merged**